File: R_shallow/DES.R
library(pracma)
library(comprehenr)
library("ringbuffer")
source("Classes.R")

is_empty <- function(x) return(length(x) ==0 )

DES <- setRefClass(
    "DES",
    contains = "Algorithm",
    fields = list(
        defaultMean = "numeric",
        lambda = "numeric",
        mu = "numeric",
        mueff = "numeric",
        cc = "numeric",
        pathLength = "numeric",
        Ft = "numeric",
        c_Ft = "numeric",
        cp = "numeric",
        weights = "numeric",
        weightsPop = "numeric",
        pathRatio = "numeric",
        histSize = "numeric"
    ),
    methods = list(
        init_default_parameters = function(){
            if(is_empty(defaultMean)){
                defaultMean <<- numeric(objective_fun$dim)
            }
            if(is_empty(population)){
                population <<- gen_random_population()
            }
            if(is_empty(lambda)){
                lambda <<- dim(population)[1]
            }
            if(is_empty(mu)){
                mu <<- floor(lambda/2)
            }
            if(is_empty(weights)){
                tmp <- log(mu+1) - log(1:mu)
                weights <<- tmp/sum(tmp)
            }
            if(is_empty(weightsPop)){
                tmp <- log(lambda+1) - log(1:lambda)
                weightsPop <<- tmp/sum(tmp)
            }
            if(is_empty(mueff)){
                mueff <<- sum(weights)^2/sum(weights^2)
            }
            if(is_empty(cc)){
                cc <<- mu/(mu+2)
            }
            if(is_empty(pathLength)){
                pathLength <<- 6
            }
            if(is_empty(cp)){
                cp <<- 1/sqrt(objective_fun$dim)
            }
            if(is_empty(Ft)){
                Ft <<- 1
            }
            if(is_empty(c_Ft)){
                c_Ft <<- 0
            }
            if(is_empty(pathRatio)){
                pathRatio <<- sqrt(pathLength)
            }
            if(is_empty(histSize)){
                histSize <<- ceiling(6+ceiling(3*sqrt(objective_fun$dim)))
            }
        },

        population_2_matrix = function(df){
            return(t(data.matrix(df[,!(names(df)=='fitness')])))
        },

        matrix_2_population = function(mat){
            fitness <- c()
            for(i in 1:dim(mat)[2]){
                fitness[i]<-objective_fun$evaluate(mat[,i])
            }
            return(data.frame(t(mat), fitness))
        },

        selection = function(df, offspring_num){
            df_sorted <- df[order(df['fitness'], decreasing=FALSE),]
            selected_points <- df_sorted[1:offspring_num,]
            return(selected_points)
        },

        # sampleFromHistory = function(history,historySample,lambda){
        # # ret <- c()
        # # for(i in 1:lambda)
        # #     # ret <- c(ret,sample(1:ncol(history[[historySample[i]]]), 1))
        # #     ret <- c(ret,sample(1:mu, 1))
        # ret <- sample(1:mu,lambda,T)
        # return(ret)
        # },

        run = function(){
            # default parameter init
            init_default_parameters()

            if(point_dim == 2){
                LOG_POPS <- TRUE
                first_pops <<- list(population) 
            }else{
                LOG_POPS <- FALSE
            }

            tol <- 10e-30
            # dynamic parameter init
            # steps       <- ringbuffer(size = (pathLength*objective_fun$dim))   
            dMean       <- array(0, dim=c(objective_fun$dim,histSize))
            # FtHistory   <- array(0, histSize)                                 
            pc       <- array(0, dim=c(objective_fun$dim,histSize))

            # history init
            histHead    <- 0
            history     <- list()

            # population <<- gen_random_population()
            population_mat <- population_2_matrix(population)
            best_point <- sel_best()

            oldMean         <- numeric(objective_fun$dim)
            newMean         <- defaultMean
  
            # Store population and selection means
            # print(population_mat)
            # print(weightsPop)
            popMean         <- drop(population_mat %*% weightsPop)
            muMean          <- newMean

            ## Matrices for creating diffs
            diffs     <- matrix(0, objective_fun$dim, lambda)
            x1sample  <- numeric(lambda)
            x2sample  <- numeric(lambda)

            # chiN      <- (sqrt(2)*gamma((objective_fun$dim+1)/2)/gamma(objective_fun$dim/2))
            histNorm  <- 1/sqrt(2)

            mean_time <- 0
            iterations  <<- 0
            times_vec <- c()
            while(iterations < max_iter && objective_fun$evaluate(best_point) > tolerance){
                start_time <- Sys.time()
                if(LOG_POPS && iterations <20){
                    first_pops <<- list.append(first_pops, population)
                }

                iterations  <<- iterations + 1L
                histHead  <- (histHead %% histSize) + 1

                selectedPoints_df <- selection(population, offspring_num=mu)
                selectedPoints_mat <- population_2_matrix(selectedPoints_df)

                # Save selected population in the history buffer
                history[[histHead]] <- array(0,dim=c(objective_fun$dim,mu))
                history[[histHead]] <- selectedPoints_mat * histNorm/Ft

                ## Calculate weighted mean of selected points
                oldMean <- newMean
                newMean <- drop(selectedPoints_mat %*% weights)

                ## Write to buffers
                muMean <- newMean
                dMean[,histHead] <- (muMean - popMean) / Ft

                step <- (newMean - oldMean) / Ft
                # ## Update Ft
                # FtHistory[histHead] = Ft

                if(histHead==1){
                    pc[,histHead] = (1 - cp)* numeric(objective_fun$dim)/sqrt(objective_fun$dim) + sqrt(mu * cp * (2-cp))* step
                # pc[,histHead] = (1 - cp)* rep(0.0, objective_fun$dim)/sqrt(objective_fun$dim) + sqrt(mu * cp * (2-cp))* step
                }else{
                pc[,histHead] = (1 - cp)* pc[,histHead-1] + sqrt(mu * cp * (2-cp))* step
                }
                ## Sample from history with uniform distribution
                limit <- min(iterations, histSize)
                historySample <- sample(1:limit,lambda, T)
                historySample2 <- sample(1:limit,lambda, T)

                # x1sample <- sampleFromHistory(history,historySample,lambda)
                # x2sample <- sampleFromHistory(history,historySample,lambda)
                x1sample <- sample(1:mu, lambda, T)
                x2sample <- sample(1:mu, lambda, T)

                ## Make diffs
                for (i in 1:lambda) {
                    x1 <- history[[historySample[i]]][,x1sample[i]]
                    x2 <- history[[historySample[i]]][,x2sample[i]]

                    diffs[,i] <- sqrt(cc)*( (x1 - x2) + rnorm(1)*dMean[,historySample[i]] ) + sqrt(1-cc) * rnorm(1)*pc[,historySample2[i]]
                }

                ## New population
                # population_mat <- newMean + Ft * diffs + tol*rnorm(diffs)/chiN
                population_mat <- newMean + Ft * diffs
                # population_mat <- deleteInfsNaNs(population_mat)

                population <<- matrix_2_population(population_mat)
                best_point <- sel_best()
                popMean <- drop(population_mat %*% weightsPop)
                times_vec[iterations] <- (Sys.time() - start_time)
                # mean_time <- times_vec[iterations]
            }
        mean_time <- sum(times_vec)/iterations
        mean_time <- as.numeric(mean_time, units="secs")
        return(new('Result', best_point=best_point, end_reason='max_iter',mean_iteration_time=mean_time, iterations=iterations, times_list=times_vec))        
    }
    )
) 

File: Python_shallow/DES.py
from Algorithm import Algorithm
from Point import Point
import math
import numpy as np


class DES(Algorithm):

    def __init__(self, objective_fun=None, start_pop=None, population_filename=None, plot_data=False, time_eval=False,
                 mean=None, lbd=None, hist_size=None, path_len=None):

        Algorithm.__init__(self, objective_fun, start_pop,
                           population_filename, plot_data, time_eval)
        self.m = mean
        self.lbd = lbd
        # if lbd:
        #     self.lbd = lbd
        # else:
        #     self.lbd = 4*self.point_dim
        self.mu = None
        self.weights = None
        self.weights_pop = None
        self.mu_eff = None
        self.cc = None
        self.cp = None
        self.Ft = None
        self.c_Ft = None
        self.path_len = path_len
        self.path_ratio = None
        self.hist_size = hist_size
        if self.population and self.point_dim:
            self.init_default_parameters()

    def init_default_parameters(self):
        if not self.m:
            self.m = self.get_population_mean()
        if not self.lbd:
            self.lbd = len(self.population)
            # self.lbd = 4*self.point_dim
        if not self.mu:
            self.mu = math.floor(self.lbd/2)
        if not self.weights:
            temp = np.array([math.log(self.mu+1) - math.log(i)
                             for i in range(1, self.mu+1)])
            self.weights = temp/np.sum(temp)
        if not self.weights_pop:
            temp = np.array([math.log(self.lbd+1) - math.log(i)
                             for i in range(1, self.lbd+1)])
            self.weights_pop = temp/np.sum(temp)
        if not self.mu_eff:
            self.mu_eff = np.sum(self.weights)**2/np.sum(self.weights**2)
        if not self.cc:
            self.cc = self.mu/(self.mu + 2)
        if not self.cp:
            self.cp = 1/math.sqrt(self.point_dim)
        if not self.Ft:
            self.Ft = 1
        if not self.c_Ft:
            self.c_Ft = 0
        if not self.path_len:
            self.path_len = 6
        if not self.path_ratio:
            self.path_ratio = math.sqrt(self.path_len)
        if not self.hist_size:
            self.hist_size = math.ceil(
                6 + math.ceil(3*math.sqrt(self.point_dim)))

    @staticmethod
    def sort_by_fitness(population):
        fitnesses = np.array([p.value for p in population])
        index_order = [int(i) for i in fitnesses.argsort()]
        pop_sorted = [population[i] for i in index_order]
        return pop_sorted

    def selection(self, population):
        pop_sorted = DES.sort_by_fitness(population)
        return pop_sorted[:self.mu]

    def single_iteration(self, history, hist_norm, pop_mean, d_mean, pc, diffs):
        self.plot_population()
        best_point = self.sel_best()
        # print(best_point.value)

        hist_iterator = self.iterations % self.hist_size
        selected_points = self.selection(self.population)
        selected_points_mat = np.array(
            [p.coordinates for p in selected_points]).T

        history[hist_iterator] = selected_points_mat * hist_norm/self.Ft

        new_m = selected_points_mat @ self.weights
        d_mean[:, hist_iterator] = (new_m - pop_mean) / self.Ft

        step = (new_m - self.m) / self.Ft

        if hist_iterator == 0:
            pc[:, hist_iterator] = (1-self.cp) * np.zeros(self.point_dim)/math.sqrt(
                self.point_dim) + math.sqrt(self.mu*self.cp*(2-self.cp))*step
        else:
            pc[:, hist_iterator] = (
                1-self.cp) * pc[:, hist_iterator-1] + math.sqrt(self.mu*self.cp*(2-self.cp))*step

        limit = min(self.iterations, self.hist_size-1)

        history_sample_1 = np.random.choice(
            list(range(limit+1)), size=self.lbd)
        history_sample_2 = np.random.choice(
            list(range(limit+1)), size=self.lbd)

        x1_sample = np.random.choice(list(range(self.mu)), size=self.lbd)
        x2_sample = np.random.choice(list(range(self.mu)), size=self.lbd)

        for i in range(self.lbd):
            x1 = history[history_sample_1[i]][:, x1_sample[i]]
            x2 = history[history_sample_1[i]][:, x2_sample[i]]

            diffs[:, i] = math.sqrt(self.cc)*((x1-x2) + np.random.normal(size=1)*d_mean[:, history_sample_1[i]]
                                              ) + math.sqrt(1-self.cc) * np.random.normal(size=1)*pc[:, history_sample_2[i]]

        new_population_mat = new_m + self.Ft * diffs.T
        pop_mean = new_population_mat.T @ self.weights_pop

        new_population = [Point(coordinates=c, objective_fun=self.objective_fun)
                          for c in new_population_mat]

        prev_best = best_point
        self.population = new_population
        if self.objective_fun.bounds:
            self.population = self.objective_fun.repair_points(self.population)
        self.m = new_m

        self.iterations += 1

        return history, hist_norm, pop_mean, d_mean, pc, diffs, prev_best

    def run(self):
        assert self.point_dim is not None
        self.lbd = 4*self.point_dim
        if self.population is None and not self.lbd:
            self.gen_random_population()
        elif self.population is None and self.lbd:
            self.gen_random_population(size=self.lbd)
        if not self.is_initialized():
            self.init_default_parameters()

        d_mean = np.zeros(shape=(self.point_dim, self.hist_size))
        pc = np.zeros(shape=(self.point_dim, self.hist_size))
        history = [None]*self.hist_size
        diffs = np.zeros(shape=(self.point_dim, self.lbd))

        pop_coordinates = np.array([p.coordinates for p in self.population])
        pop_mean = pop_coordinates.T @ self.weights_pop

        hist_norm = 1/math.sqrt(2)

        prev_best = self.sel_best()
        mean_time = 0 if self.time_eval else None
        times_list = []
        self.iterations = 0

        while not self.check_end_cond(prev_best=prev_best):
            try:
                if self.time_eval:
                    payload = self.evaluate_single_iteration_with_time(
                        history, hist_norm, pop_mean, d_mean, pc, diffs)
                    history, hist_norm, pop_mean, d_mean, pc, diffs, best = payload['data']
                    mean_time += payload['time']
                    times_list.append(payload['time'])
                else:
                    history, hist_norm, pop_mean, d_mean, pc, diffs, best = self.single_iteration(
                        history, hist_norm, pop_mean, d_mean, pc, diffs)
            except ArithmeticError:
                break
            if prev_best.value >= best.value:
                prev_best = best

        if mean_time is not None:
            mean_time = mean_time/self.iterations
        return Algorithm.Result(best_point=self.sel_best(), end_reason=self.end_reason, mean_iteration_time=mean_time, iteration_num=self.iterations, times_list=times_list)
